### PR TITLE
Refactor `loop.setup_data` with utility functions

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Follow-up to https://github.com/Lightning-AI/lightning/pull/16726

This PR is pure refactoring into smaller functional utilities that we share. No intended logic changes

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/pytorch/loops/fit_loop.py [hunk 1]
```diff
@@ -209,52 +215,36 @@ def setup_data(self, shuffle: bool = True) -> None:
...
 
-        if isinstance(trainer.limit_train_batches, int):
-            self.max_batches = min(orig_train_batches, trainer.limit_train_batches)
-        elif self.max_batches != float("inf"):
-            self.max_batches = int(orig_train_batches * trainer.limit_train_batches)
-        elif trainer.limit_train_batches != 1.0:
-            raise MisconfigurationException(
-                "When using an `IterableDataset`, `Trainer(limit_train_batches)` must be `1.0` or an int."
-                "An int specifies `num_training_batches` to use."
-            )
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/fit_loop.py:256]: This logic isn't part of the other stages, but you are now calling _parse_num_batches there, which could produce additional user errors that weren't there before.

File: src/lightning/pytorch/loops/fit_loop.py [hunk 2]
```diff
@@ -286,20 +276,6 @@ def setup_data(self, shuffle: bool = True) -> None:
...
-            and isinstance(trainer.limit_train_batches, float)
-            and orig_train_batches != float("inf")
-        ):
-            min_percentage = 1.0 / orig_train_batches
-            raise MisconfigurationException(
-                f"You requested to check {trainer.limit_train_batches} of the `train_dataloader` but"
-                f" {trainer.limit_train_batches} * {orig_train_batches} < 1. Please increase the"
-                f" `limit_train_batches` argument. Try at least"
-                f" `limit_train_batches={min_percentage}`"
-            )
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/fit_loop.py:301]: Same here

File: src/lightning/pytorch/loops/prediction_loop.py [hunk 3]
```diff
@@ -114,17 +119,34 @@ def run(self) -> Optional[_PREDICT_OUTPUT]:
...
         trainer = self.trainer
         source = self._data_source
-        pl_module = trainer.lightning_module
         # a dfault `predict_step` exists in the LightningModule, so no need to check if it's overridden
         if not source.is_defined() or trainer.limit_predict_batches == 0:
             return
 
-        self.max_batches, combined_loader = trainer._data_connector._reset_eval_dataloader(
-            RunningStage.PREDICTING, model=pl_module
-        )
```
Comment (awaelchli) [inline src/lightning/pytorch/loops/prediction_loop.py:124]: isn't there a resolve_overfit_batches missing somewhere here? It was part of the _reset_eval_dataloader, and as I understand, the code was partially moved here.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/loops/prediction_loop.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/pytorch/loops/prediction_loop.py:124]: You are correct that the overfit batches check was run before and not anymore. However, I don't see any source mentioning that `overfit_batches` is meant to work with `trainer.predict()`. Do you think it should? What would be the benefit?

Comment (carmocca) [reply src/lightning/pytorch/loops/prediction_loop.py:124]: Actually, we were doing it unnecessarily for `trainer.validate/test`: [`9320d34` (#16918)](https://github.com/Lightning-AI/lightning/pull/16918/commits/9320d34cddea954664d803eba80dd44bc2cf96b4)

**Review by justusschock — APPROVED** [review]

**Pull request merged**